src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs

```diff
@@ -0,0 +1,90 @@
+
+        /// <summary>
+        /// Constructs DefaultJsonTypeInfoResolver.
+        /// </summary>
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+        public DefaultJsonTypeInfoResolver() : this(mutable: true)
+        {
+        }
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+        internal DefaultJsonTypeInfoResolver(bool mutable)
+        {
+            Modifiers = new ConfigurationList<Action<JsonTypeInfo>>() { VerifyMutable = VerifyMutable };
+            _mutable = mutable;
+        }
+
+        /// <inheritdoc/>
+        public JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
+        {
+            _mutable = false;
+
+            JsonTypeInfo.ValidateType(type, null, null, options);
+            JsonTypeInfo typeInfo = CreateJsonTypeInfo(type, options);
+
+            foreach (Action<JsonTypeInfo> modifier in Modifiers)
+            {
+                modifier(typeInfo);
+            }
+
+            return typeInfo;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "The ctor is marked RequiresUnreferencedCode.")]
+        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
+            Justification = "The ctor is marked RequiresDynamicCode.")]
+        private JsonTypeInfo CreateJsonTypeInfo(Type type, JsonSerializerOptions options)
+        {
+            MethodInfo methodInfo = typeof(JsonSerializerOptions).GetMethod(nameof(JsonSerializerOptions.CreateReflectionJsonTypeInfo), BindingFlags.NonPublic | BindingFlags.Instance)!;
+#if NETCOREAPP
+            return (JsonTypeInfo)methodInfo.MakeGenericMethod(type).Invoke(options, BindingFlags.NonPublic | BindingFlags.DoNotWrapExceptions, null, null, null)!;
+#else
+            try
+            {
+                return (JsonTypeInfo)methodInfo.MakeGenericMethod(type).Invoke(options, null)!;
+            }
+            catch (TargetInvocationException ex)
+            {
+                // Some of the validation is done during construction (i.e. validity of JsonConverter, inner types etc.)
+                // therefore we need to unwrap TargetInvocationException for better user experience
+                ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
+                throw null!;
+            }
+#endif
+        }
+
+        /// <summary>
+        /// List of JsonTypeInfo modifiers. Modifying callbacks are called consecutively after initial resolution
+        /// and cannot be changed after GetTypeInfo is called.
+        /// </summary>
+        public IList<Action<JsonTypeInfo>> Modifiers { get; }
+
+        private void VerifyMutable()
+        {
+            if (!_mutable)
```

Comment:
All new functionality should include testing in this PR. It's very likely that we might forget adding coverage for some of these components if we regard testing as an afterthought.

Reply:
I add tests per every API added or behavior change. I don't see how I could omit it. It just makes for less noise in the final PR and makes it easier to spot any testing gaps